crimson/osd: execute `submit_error_log` as an ExclusivePhase

Matan-B opened the pull request:
~Blocked by: #54040~

Introduced: https://github.com/ceph/ceph/pull/39772

Fixes: https://tracker.ceph.com/issues/61651


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

Matan-B commented:
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-do_osd_ops_execute-v3/

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -874,38 +874,75 @@ PG::do_osd_ops_execute(
+
+          TODO: 1) Is it possible to handle `pre_all_completed_fut`
+                   the errors before actaully getting here?
+
```
I'm not entirely sure I'm following your intent here, but I think the problem is that splitting error_log_fut from submitted_fut doesn't help and actually makes this more complicated.  Ultimately, both error_log_fut and submitted_fut encode the same thing -- that we've fully updated the in-memory PG state, submitted (ordered) the resulting messages, and are ready to process a new operation.  Whether this was via the normal repop pathway or the error log pathway doesn't matter.  I don't think the caller cares about the difference.

I think this probably means that the submit_error_log call should be moved out of do_osd_ops into do_osd_ops_execute.  The submitted_fut should encode the normal repop path if there are no errors and the submit_error_log path if there are.

I think the failure_func argument to do_osd_ops_execute is probably useless at that point.  If the caller needs to know, the error value can be embedded into either all_completed_fut or submitted_fut (not as an error -- just as a normal value).

Matan-B commented:
@athanatos, the call to error_log_fut was moved to do_osd_ops_execute and the original split between do_osd_ops_execute and submitted_fut is reverted. The diff is much cleaner now. However, IIUC, this is still doesn't solve the issue described in b02676b60515667592240486a8a5695f9d74fbbd.

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -903,23 +921,26 @@ PG::do_osd_ops_execute(
-    );
+    return maybe_rollback_fut.then_interruptible(
+    [error_func_ptr, e, rep_tid, failure_func_ptr] {
+      return (*error_func_ptr)(e, rep_tid).then(
+      [failure_func_ptr, e, rep_tid] {
+        return PG::do_osd_ops_iertr::make_ready_future<pg_rep_op_fut_t<Ret>>(
+          std::move(seastar::now()),
+          std::move((*failure_func_ptr)(e, rep_tid))
```
1/2: This part lets `submit_error_log` run from within `do_osd_ops_execute` which resolves the original issue.

athanatos replied:
This one is chained after the future returned from submit_transaction, which would otherwise resolve into the submitted and completed futures.  This one, I think, as you observe is the one that resolves the bug.

athanatos replied:
It seems to me that error_func_ptr could return an eversion_t, pass it into failure_func_ptr, and thereby avoid the log_entry_version map.

If we retain the log_entry_version map, it probably needs to be cleared during interval change.

Matan-B replied:
> It seems to me that error_func_ptr could return an eversion_t, pass it into failure_func_ptr, and thereby avoid the log_entry_version map.

After the last commit, it is no longer possible.

> If we retain the log_entry_version map, it probably needs to be cleared during interval change.
 
Fixed.

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -890,11 +904,15 @@ PG::do_osd_ops_execute(
           return rollbacker.rollback_obc_if_modified(e).then_interruptible(
-          [this, e, failure_func_ptr] {
-            return (*failure_func_ptr)(e , shard_services.get_tid());
+          [this, e, failure_func_ptr, error_func_ptr] {
+            auto rep_tid = shard_services.get_tid();
+            return (*error_func_ptr)(e, rep_tid)
+            .then([failure_func_ptr, e, rep_tid] {
+              return (*failure_func_ptr)(e , rep_tid);
+            });
```
2/2: The second `submit_error_log` call, since it's being run from within `do_osd_ops_execute` it should also be safe . Even though it's packed inside `all_completed_fut` below (in case of an error), it will be executed before returning `pg_rep_op_fut_t` to the caller.

athanatos replied:
Hmm, this handler is actually chained after _all_complete_fut and could therefore happen after we leave the exclusive phase.  On the other hand, I think this case is actually impossible -- any error here would have to come from the _all_completed_fut returned from submit_transaction, which would have to be an error with the actual transaction submitted to the backing store or somehow with the messages received from peers.  I *think* any error here could be treated as fatal to the OSD.

Matan-B replied:
> Hmm, this handler is actually chained after _all_complete_fut and could therefore happen after we leave the exclusive phase. On the other hand, I think this case is actually impossible -- any error here would have to come from the _all_completed_fut returned from submit_transaction, which would have to be an error with the actual transaction submitted to the backing store or somehow with the messages received from peers. I _think_ any error here could be treated as fatal to the OSD.

This case was possible since there were non-fatal errors that could have been returned as well. I added an assert to verify that we handle only those.
Moreover, I moved the call to error_log_ptr earlier - before actually returning the non-fatal errors. That way, it will be executed during the exclusive phase.

Matan-B commented:
https://pulpito.ceph.com/matan-2023-11-08_11:46:17-crimson-rados-wip-matanb-crimson-do_osd_ops_execute-v3-distro-crimson-smithi/

Matan-B commented:
https://pulpito.ceph.com/matan-2023-11-08_11:46:17-crimson-rados-wip-matanb-crimson-do_osd_ops_execute-v3-distro-crimson-smithi/

30/34 passed.
Results looks good (!)
The issue was previously seen consistently in the suite and seems to be resolved.

Failures look unrelated and are new since they were previously hidden by the issue resolved in this PR:
* 7451909 - osd.3 crash `SnapTrimObjSubEvent` lifetime
* 7451911 - osd.1 and osd.2  -  BlueStore::omap_get_values (Addresed: https://github.com/ceph/ceph/pull/54214/commits/7b8795aa28c6c599cf0b0ac546b958c635a9dd0a) 
* 7451915 - osd.3 crash PG lifetime (Addresed: #54416)
* 7451923 - [  FAILED  ] TestClsRbd.directory_methods

athanatos reviewed (changes requested):
I think there's further simplification that can be done, but we could merge this version as a fix.  If we go that route, I'd just request that the log_entry_version map be cleared on interval change.

Matan-B commented:
> I think there's further simplification that can be done, but we could merge this version as a fix. If we go that route, I'd just request that the log_entry_version map be cleared on interval change.

@athanatos, I've addressed the comments as new commits for easier CR.

Matan-B commented:
Test results look good:
https://pulpito.ceph.com/matan-2023-11-16_08:02:56-crimson-rados-wip-matanb-crimson-testing-11-15-distro-crimson-smithi/

Unrelated:
7460428 - https://tracker.ceph.com/issues/63556
7460429 - https://tracker.ceph.com/issues/62162
7460443 - https://tracker.ceph.com/issues/62740
7460420 - Addressed: https://github.com/ceph/ceph/pull/54513

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -850,23 +851,29 @@ PG::do_osd_ops_execute(
+        }
+      } else {
+        // drop request
+        return (*error_func_ptr)(EAGAIN, rep_tid).then([] {
```
We shouldn't be calling submit_error_log in these cases.

submit_error_log records the result of an IO into the pg log so that we can return the same error code if the client resends the request.  This is *only relevant* for logical errors resulting from the target object state -- for example, EEXIST returned on an exclusive create -- because there is application logic built to rely on them.  In classic, the only such site is if the return value from do_osd_ops is negative (or the transaction is empty) -- see PrimaryLogPG::prepare_transaction, specifically where we set update_log_only to true.

We do not want to record space usage errors or errors specific to conditions on the primary OSD such as IO errors.  submit_error_log isn't a catch-all error path.

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -328,6 +329,12 @@ void PG::on_removal(ceph::os::Transaction &t) {
   }));
 }
 
+void PG::cancel_log_updates()
```
clear_log_entry_maps

Matan-B commented:
https://pulpito.ceph.com/matan-2023-11-19_17:04:13-crimson-rados-wip-matanb-crimson-testing-11-19-distro-crimson-smithi/

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -318,6 +318,7 @@ unsigned PG::get_target_pg_log_entries() const
 }
 
 void PG::on_removal(ceph::os::Transaction &t) {
+  clear_log_entry_maps();
```
This shouldn't be necessary -- any PG invoking on_removal would have to have gone through on_change since the last time it was primary and placed anything in the map.  You can assert here if you like, but on_change should be the only place this needs to be cleared.

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -873,7 +894,7 @@ PG::do_osd_ops_execute(
     });
   }).safe_then_unpack_interruptible(
-    [success_func=std::move(success_func), rollbacker, this, failure_func_ptr]
+    [success_func=std::move(success_func), error_func_ptr, rollbacker, this, failure_func_ptr]
```
This capture seems to be unused.

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -905,32 +931,39 @@ PG::do_osd_ops_execute(
+    return maybe_rollback_fut.then_interruptible(
+    [error_func_ptr, e, rep_tid, failure_func_ptr] {
+      // record error log
+      return (*error_func_ptr)(e, rep_tid).then(
```
This appears to be the only actual user of error_func_ptr -- why not simply move the declaration of maybe_submit_error_log here and avoid needing error_func_ptr entirely?

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -1036,45 +1085,41 @@ PG::do_osd_ops(
+        peering_state.complete_write(log_entry_version[rep_tid], last_complete);
+        log_entry_version.erase(rep_tid);
+        logger().debug("do_osd_ops_execute::failure_func write complete,"
+                        " erasing rep_tid {}", rep_tid);
```
This else branch seems like it should have been moved into do_osd_ops_execute as well.  I'd probably factor it into a helper (complete_error_log) and chain it after the error_func_ptr invocation.

Matan-B replied:
#54765

athanatos reviewed (approved):
There's more simplification to be done, but I'd be ok with doing it in a subsequent PR if this one is passing tests.  Up to you.

Matan-B commented:
> There's more simplification to be done, but I'd be ok with doing it in a subsequent PR if this one is passing tests. Up to you.

I prefer addressing the cleanups in a subsequent PR to let main branch not to have this issue anymore. Thanks!

cyx1231st commented on src/crimson/osd/pg.cc:
```diff
@@ -910,23 +924,27 @@ PG::do_osd_ops_execute(
+    return maybe_rollback_fut.then_interruptible(
+    [error_func_ptr, e, rep_tid, failure_func_ptr] {
+      // record error log
+      return (*error_func_ptr)(e, rep_tid).then(
```
Ack, `submit_error_log()` is moved from the concurrent wait_repop phase to the exclusive process phase, which should address the ordering issue.

Matan-B replied:
Not encountered since merge _yet_ :)
